### PR TITLE
Automatically set CHPL_GPU_CODEGEN if not already set

### PR DESCRIPTION
**Pull request by stonea**
This change automatically sets `CHPL_GPU_CODEGEN` to either `cuda` or `rocm` based on whether the user has `nvcc` or `hipcc` in their path. If they have both or neither then this variable must be set manually.

This PR also adds in some checks when `CHPL_GPU_CODEGEN=rocm` to check that the user's LLVM installation supports the AMDGPU target. If it can't validate this it suggests using the bundled llvm.

### CODE REVIEW

File: util/chplenv/chpl_gpu.py [hunk 1]
```diff
@@ -15,6 +15,19 @@
...
+    typeFound = None
+    for gpuType in GPU_TYPES.keys():
+        if which(GPU_TYPES[gpuType][1]):
+            typeFound = gpuType
+            gpuTypesFound += 1
+
+    if gpuTypesFound != 1:
+        error("Unable to determine GPU type, assign 'CHPL_GPU_CODEGEN' to one of {}".
+            format(GPU_TYPES.keys()))
+    return typeFound
```
Comment (DanilaFe) [inline util/chplenv/chpl_gpu.py:29]: Nitpicky: you could use a `gpuTypes = []` array, push to it each time `which` succeeds, and have the error message report which supported types were found. Then, no need for a counter (you've got `len`!) and the error message is a bit friendlier.

File: util/chplenv/chpl_gpu.py [hunk 2]
```diff
@@ -114,5 +127,24 @@ def get_runtime():
...
+
+@memoize
 def validate(chplLocaleModel, chplComm):
-    pass
+    if chplLocaleModel != "gpu":
+        return True
+
+    if get() == 'rocm':
+        if not validateLlvmBuiltForRocm():
+            error("LLVM not built for AMDGPU, consider setting CHPL_LLVM to 'bundled'")
```
Comment (DanilaFe) [inline util/chplenv/chpl_gpu.py:148]: Don't we also use LLVM's cuda target? Perhaps we should also validate that the LLVM is built for cuda when that's in use? Seems to be easy enough to generalize `validateLlvmBuiltForRocm`.

Comment (stonea) [reply util/chplenv/chpl_gpu.py:148]: Good point, I'll add it a check for that

File: util/chplenv/chpl_gpu.py (hunk 1 above)
Comment (DanilaFe) [reply util/chplenv/chpl_gpu.py:29]: I'd keep "one of {}" was before, but maybe add a parenthetical ("detected bla, bla SDKs") like this.

Comment (stonea) [reply util/chplenv/chpl_gpu.py]: Sure, updated in 0493a4bdef99245b760c4d2929a553ef2eeebeb1 to add a parenthetical if any gpuTypes were detected.

File: util/chplenv/chpl_gpu.py (hunk 2 above)
Comment (stonea) [reply util/chplenv/chpl_gpu.py]: Updated in 0493a4bdef99245b760c4d2929a553ef2eeebeb1

**Review by DanilaFe — APPROVED** [review]
Looks great, thanks!

**Comment by bradcray** [comment]
Nice!

**Pull request merged**